pkg/autodiscovery/listeners/kube_endpoints.go

```diff
@@ -0,0 +1,336 @@
+			if first.Subsets[i].Addresses[j].IP != second.Subsets[i].Addresses[j].IP {
+				return true
+			}
+			if first.Subsets[i].Addresses[j].Hostname != second.Subsets[i].Addresses[j].Hostname {
+				return true
+			}
+		}
+		// Ports
+		for j := range first.Subsets[i].Ports {
+			if first.Subsets[i].Ports[j].Port != second.Subsets[i].Ports[j].Port {
+				return true
+			}
+			if first.Subsets[i].Ports[j].Name != second.Subsets[i].Ports[j].Name {
+				return true
+			}
+		}
+	}
+	// No relevant change
+	return false
+}
+
+// isEndpointsAnnotated looks for the corresponding service of a kubernetes endpoints object
+// and returns true if the service has endpoints annotations, otherwise returns false.
+func (l *KubeEndpointsListener) isEndpointsAnnotated(kep *v1.Endpoints) bool {
+	ksvc, err := l.serviceLister.Services(kep.Namespace).Get(kep.Name)
+	if err != nil {
+		log.Debugf("Cannot get Kubernetes service: %s", err)
+	}
+	if ksvc != nil {
+		if _, found := ksvc.Annotations[kubeEndpointsAnnotationFormat]; found {
+			return true
+		}
+	}
+	return false
+}
+
+func (l *KubeEndpointsListener) createService(kep *v1.Endpoints, firstRun bool) {
+	if kep == nil {
+		return
+	}
+	if !l.isEndpointsAnnotated(kep) {
+		// Ignore endpoints with no AD annotation on their corresponding service
+		return
+	}
+
+	eps := processEndpoints(kep, firstRun)
+
+	l.m.Lock()
+	l.endpoints[kep.UID] = eps
+	l.m.Unlock()
+
+	for _, ep := range eps {
+		l.newService <- ep
+	}
+}
+
+// processEndpoints parses a kubernetes Endpoints object
+// and returns a slice of KubeEndpointService per endpoint
+func processEndpoints(kep *v1.Endpoints, firstRun bool) []*KubeEndpointService {
+	var eps []*KubeEndpointService
+	for i := range kep.Subsets {
+		ports := []ContainerPort{}
+		// Ports
+		for _, port := range kep.Subsets[i].Ports {
+			ports = append(ports, ContainerPort{int(port.Port), port.Name})
+		}
+		// Hosts
+		for _, host := range kep.Subsets[i].Addresses {
+			// create a separate AD service per host
+			ep := &KubeEndpointService{
+				entity:       apiserver.EntityForEndpoints(kep.Namespace, kep.Name, host.IP),
+				creationTime: integration.After,
+				hosts:        map[string]string{"endpoint": host.IP},
+				ports:        ports,
+				tags: []string{
+					fmt.Sprintf("kube_service:%s", kep.Name),
+					fmt.Sprintf("kube_namespace:%s", kep.Namespace),
+					fmt.Sprintf("kube_endpoint_ip:%s", host.IP),
+				},
+			}
+			if firstRun {
+				ep.creationTime = integration.Before
+			}
+			eps = append(eps, ep)
+		}
+	}
+	return eps
+}
+
+func (l *KubeEndpointsListener) removeService(kep *v1.Endpoints) {
+	if kep == nil {
+		return
+	}
+	l.m.RLock()
+	eps, ok := l.endpoints[kep.UID]
+	l.m.RUnlock()
+	if ok {
+		l.m.Lock()
+		delete(l.endpoints, kep.UID)
+		l.m.Unlock()
+		for _, ep := range eps {
+			l.delService <- ep
+		}
+	} else {
+		log.Debugf("Entity %s not found, not removing", kep.UID)
+	}
+}
+
+// GetEntity returns the unique entity name linked to that service
```

Comment:
we create a `KubeEndpointService` this way:
```
ep := &KubeEndpointService{
				entity:       apiserver.EntityForEndpoints(kep.Namespace, kep.Name, host.IP),
				creationTime: integration.After,
				hosts:        map[string]string{"endpoint": host.IP},
				ports:        ports,
				tags: []string{
					fmt.Sprintf("kube_service:%s", kep.Name),
					fmt.Sprintf("kube_namespace:%s", kep.Namespace),
					fmt.Sprintf("kube_endpoint_ip:%s", host.IP),
				},
			}
```
I don't see how it could happen to have nil attributes (`ports` is already initialized `ports := []ContainerPort{}`)

Reply:
Yes, but as they are public functions they could be reused by someone else who might not know about the correct way to init the  `KubeEndpointService `.
I just know it's a good practice to check for nil pointer where they are accessed

Reply:
right 👍 I'll check for nil pointers we `return`